README.md
# @neonox31/node-red-contrib-zigate

This repository contains a set of [node-red](https://nodered.org/) nodes to interract with [zigate](https://zigate.fr) gateway.

All nodes depends of my personnal typescript library located [here](https://github.com/Neonox31/zigate)

⚠ **This is not a zigate official repo**

## Install

### Use node-red palette
Simply search for `@neonox31/node-red-contrib-zigate` and press `install` button.

### Manually
Use `npm i @neonox31/node-red-contrib-zigate` or `yarn add @neonox31/node-red-contrib-zigate` inside `.node-red` directory.

## Nodes

### zigate-config
This node is a node-red [configuration node](https://nodered.org/docs/creating-nodes/config-nodes) and permits to share zigate configuration between all other ziagte nodes in all flows.

Currently, only one **mandatory** parameter is available which is the serial port of the plugged zigate USB dongle :
![](img/config-node-properties.png)

### zigate-in

This node listens to all zigate messages and transmit them as a node-red message payload.
The zigate message code is also transmitted as node-red message topic.

#### Basic implementation
![](img/zigate-in-example.png)

```
[{"id":"e183e7fe.821428","type":"zigate-in","z":"e4ab82f1.ccd66","name":"","zigate":"dbb094c2.e34978","x":430,"y":300,"wires":[["fb12d3f6.f1083"]]},{"id":"fb12d3f6.f1083","type":"debug","z":"e4ab82f1.ccd66","name":"","active":true,"tosidebar":true,"console":false,"tostatus":false,"complete":"false","x":610,"y":300,"wires":[]},{"id":"dbb094c2.e34978","type":"zigate-config","z":"","serialPort":"/dev/ttyUSB0"}]
```

#### Payload example
![](img/zigate-in-example-msg.png)

### zigate-out

This node is used to send commands to zigate.

The zigate message code should be specified as node-red message topic **with decimal format**.
The node-red payload is optional and could be used to override default command parameters like `interval` for `permit join` command for example. 
(Available commands are located [here](https://github.com/Neonox31/zigate/tree/master/src/commands))

#### Basic implementation
![](img/zigate-out-example.png)

```
[{"id":"f1b301d.3652","type":"zigate-out","z":"e4ab82f1.ccd66","name":"","zigate":"dbb094c2.e34978","x":660,"y":460,"wires":[]},{"id":"e357bf4f.b90d5","type":"inject","z":"e4ab82f1.ccd66","name":"Permit join","topic":"73","payload":"{\"interval\":30}","payloadType":"json","repeat":"","crontab":"","once":false,"onceDelay":0.1,"x":480,"y":460,"wires":[["f1b301d.3652"]]},{"id":"dbb094c2.e34978","type":"zigate-config","z":"","serialPort":"/dev/ttyUSB0"}]
```

#### Message example
![](img/zigate-out-example-msg.png)

### zigate-xiaomi-aqara-weather-sensor
![](img/zigate-xiaomi-aqara-weather-sensor.png)

This node retrieves data from Xiaomi Aqara Temperature Humidity Sensor using zigate.

This node is a device node, so you **should specify the zigate device short address** in its properties:

![](img/zigate-xiaomi-aqara-weather-sensor-node-properties.png)

#### Basic implementation
![](img/zigate-xiaomi-aqara-weather-sensor-example.png)

```
[{"id":"27c7dcfc.507854","type":"zigate-xiaomi-aqara-weather-sensor","z":"245ccd2c.8c7e52","zigate":"dbb094c2.e34978","name":"","shortAddress":"4fb6","x":190,"y":100,"wires":[["6d5acc1e.924474"]]},{"id":"6d5acc1e.924474","type":"debug","z":"245ccd2c.8c7e52","name":"","active":true,"tosidebar":true,"console":false,"tostatus":false,"complete":"false","x":434,"y":100,"wires":[]},{"id":"dbb094c2.e34978","type":"zigate-config","z":"","serialPort":"/dev/ttyUSB0"}]
```

#### Payload examples
Each new measure is provided as a separate node-red message :

![](img/zigate-xiaomi-aqara-weather-sensor-example-msg-temperature.png)

![](img/zigate-xiaomi-aqara-weather-sensor-example-msg-humidity.png)

![](img/zigate-xiaomi-aqara-weather-sensor-example-msg-pressure.png)


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

<a name="1.0.1"></a>
## [1.0.1](https://github.com/Neonox31/node-red-contrib-zigate/compare/1.0.0...1.0.1) (2018-07-27)


### Bug Fixes

* **package:** add missing zigate dependency ([6c05e26](https://github.com/Neonox31/node-red-contrib-zigate/commit/6c05e26))



<a name="1.0.0"></a>
# 1.0.0 (2018-07-27)


test/in_spec.js
const helper = require("node-red-node-test-helper");
const inNode = require("../src/nodes/in/in.js");

helper.init(require.resolve('node-red'));

describe('ZiGate In Node', function () {

    beforeEach(function (done) {
        helper.startServer(done);
    });

    afterEach(function (done) {
        helper.unload();
        helper.stopServer(done);
    });

    xit('should be loaded', function (done) {
        const flow = [{ id: "n1", type: "zigate-in", name: "zigate-in" }];
        helper.load(inNode, flow, function () {
            const n1 = helper.getNode("n1");
            console.log(n1);
            n1.should.have.property('name', 'zigate-in');
            done();
        });
    });

});

src/nodes/xiaomi/aqara-button/aqara-button.js
const {catchError} = require('rxjs/operators');
const {ZGDeviceType} = require('@neonox31/zigate');
const {toNodeError} = require('../../../utils');

module.exports = function (RED) {

    function ZGXiaomiAqaraButtonNode(config) {
        RED.nodes.createNode(this, config);
        const node = this;
        const configNode = RED.nodes.getNode(config.zigate);

        if (configNode.zigate) {
            const button = configNode.zigate.createDevice(ZGDeviceType.XiaomiAqaraButton, config.shortAddress);

            button.pushes$
                .pipe(catchError(toNodeError(node)))
                .subscribe((pushes) => {
                    node.send({
                        payload: {
                            pushes
                        }
                    })
                });

            button.battery$
                .pipe(catchError(toNodeError(node)))
                .subscribe((battery) => {
                    let color = 'red';

                    if (battery.level > 70) {
                        color = 'green';
                    } else if (battery.level < 50 && battery.level > 20) {
                        color = 'yellow';
                    }

                    node.status({
                        fill: color,
                        shape: "dot",
                        text: `Battery: ${ battery.level }% (${ battery.voltage } V)`
                    });
                    node.send({
                        payload: {
                            battery
                        }
                    })
                })
        }
    }

    RED.nodes.registerType("zigate-xiaomi-aqara-button", ZGXiaomiAqaraButtonNode);

};

src/nodes/xiaomi/aqara-door-sensor/aqara-door-sensor.js
const {catchError} = require('rxjs/operators');
const {ZGDeviceType} = require('@neonox31/zigate');
const {toNodeError} = require('../../../utils');

module.exports = function (RED) {

    function ZGXiaomiAqaraDoorSensorNode(config) {
        RED.nodes.createNode(this, config);
        const node = this;
        const configNode = RED.nodes.getNode(config.zigate);

        if (configNode.zigate) {
            const doorSensor = configNode.zigate.createDevice(ZGDeviceType.XiaomiAqaraDoorSensor, config.shortAddress);

            doorSensor.state$
                .pipe(catchError(toNodeError(node)))
                .subscribe((state) => {
                    node.send({
                        payload: state
                    })
                });

            doorSensor.battery$
                .pipe(catchError(toNodeError(node)))
                .subscribe((battery) => {
                    let color = 'red';

                    if (battery.level > 70) {
                        color = 'green';
                    } else if (battery.level < 50 && battery.level > 20) {
                        color = 'yellow';
                    }

                    node.status({
                        fill: color,
                        shape: "dot",
                        text: `Battery: ${ battery.level }% (${ battery.voltage } V)`
                    });
                    node.send({
                        payload: {
                            battery
                        }
                    })
                })
        }
    }

    RED.nodes.registerType("zigate-xiaomi-aqara-door-sensor", ZGXiaomiAqaraDoorSensorNode);

};

src/nodes/xiaomi/aqara-weather-sensor/aqara-weather-sensor.js
const {catchError} = require('rxjs/operators');
const {ZGDeviceType} = require('@neonox31/zigate');
const {toNodeError} = require('../../../utils');

module.exports = function (RED) {

    function ZGXiaomiAqaraWeatherSensorNode(config) {
        RED.nodes.createNode(this, config);
        const node = this;
        const configNode = RED.nodes.getNode(config.zigate);

        if (configNode.zigate) {
            const weatherSensor = configNode.zigate.createDevice(ZGDeviceType.XiaomiAqaraWeatherSensor, config.shortAddress);

            weatherSensor.temperature$
                .pipe(catchError(toNodeError(node)))
                .subscribe((temperature) => {
                    node.send({
                        payload: {
                            temperature
                        }
                    })
                });

            weatherSensor.humidity$
                .pipe(catchError(toNodeError(node)))
                .subscribe((humidity) => {
                    node.send({
                        payload: {
                            humidity
                        }
                    })
                });

            weatherSensor.pressure$
                .pipe(catchError(toNodeError(node)))
                .subscribe((pressure) => {
                    node.send({
                        payload: {
                            pressure
                        }
                    })
                });

            weatherSensor.battery$
                .pipe(catchError(toNodeError(node)))
                .subscribe((battery) => {
                    let color = 'red';

                    if (battery.level > 70) {
                        color = 'green';
                    } else if (battery.level < 50 && battery.level > 20) {
                        color = 'yellow';
                    }

                    node.status({
                        fill: color,
                        shape: "dot",
                        text: `Battery: ${ battery.level }% (${ battery.voltage } V)`
                    });
                    node.send({
                        payload: {
                            battery
                        }
                    })
                })
        }
    }

    RED.nodes.registerType("zigate-xiaomi-aqara-weather-sensor", ZGXiaomiAqaraWeatherSensorNode);

};